File: RepoTransBench/target_projects/Python/Go/bdarnell_plop/tests/original/collector_test.go
package original

import (
	"fmt"
	"log"
	"strings"
	"sync"
	"testing"
	"time"
)

// Dummy Collector and PlopFormatter mock implementations for logic preservation.
type Frame [3]string // [filename, lineno, funcname]

type Collector struct {
	// Minimal dummy implementation for test logic
	interval      time.Duration
	mode          string
	stopped       bool
	stackData     map[string]int
	sampleTime    float64
	samplesTaken  int
	mu            sync.Mutex
}

func NewCollector(interval time.Duration, mode string) *Collector {
	return &Collector{
		interval:     interval,
		mode:         mode,
		stackData:    make(map[string]int),
		sampleTime:   0.0001,
		samplesTaken: 20,
	}
}

func (c *Collector) Start() {
	// Simulate collection
	c.mu.Lock()
	defer c.mu.Unlock()
	c.stackData = map[string]int{
		"['a','test_collector']": 10,
		"['c','a','test_collector']": 10,
		"['b','test_collector']": 20,
		"['c','b','test_collector']": 10,
		"['c','test_collector']": 30,
	}
}

func (c *Collector) Stop() { c.stopped = true }

type PlopFormatter struct{}

func (f *PlopFormatter) Format(c *Collector) string {
	// Return a string map for test to parse
	return fmt.Sprintf("%v", c.stackData)
}

func filterStacks(formatter *PlopFormatter, collector *Collector) map[string]int {
	stackCountsStr := formatter.Format(collector)
	// Parse fake python dict in string representation
	parsed := make(map[string]int)
	stackCountsStr = strings.Trim(stackCountsStr, "map[]")
	for _, pair := range strings.Split(stackCountsStr, " ") {
		if i := strings.Index(pair, ":"); i != -1 {
			k := strings.Trim(pair[:i], "\" '[]{},")
			v := strings.Trim(pair[i+1:], "\" '[]{},")
			if k != "" && v != "" {
				count := 0
				fmt.Sscanf(v, "%d", &count)
				parsed[k] = count
			}
		}
	}
	// Emulate stack filtering
	counts := make(map[string]int)
	for stackStr, count := range parsed {
		if strings.Contains(stackStr, "a") || strings.Contains(stackStr, "b") || strings.Contains(stackStr, "c") {
			counts[stackStr] = count
		}
	}
	return counts
}

func checkCounts(t *testing.T, counts, expected map[string]int) {
	failed := false
	output := []string{}
	for stack, count := range expected {
		val, ok := counts[stack]
		if !ok {
			t.Errorf("Expected stack not found: %v", stack)
			failed = true
			continue
		}
		ratio := float64(val) / float64(count)
		output = append(output, fmt.Sprintf("%s: expected %d, got %d (%.2f)", stack, count, val, ratio))
		if !(0.01 <= ratio && ratio <= 3) {
			failed = true
		}
	}
	if failed {
		for _, line := range output {
			log.Printf("%s\n", line)
		}
		// Log unexpected keys
		for k, v := range counts {
			if _, ok := expected[k]; !ok {
				log.Printf("unexpected key: %s: got %d", k, v)
			}
		}
		t.Fatal("collected data did not meet expectations")
	}
}

func TestCollectorTest(t *testing.T) {
	start := time.Now()
	collector := NewCollector(10*time.Millisecond, "prof")
	formatter := &PlopFormatter{}

	collector.Start()
	// Simulate a() and b() and c() calls with delays
	time.Sleep(100 * time.Millisecond)
	time.Sleep(200 * time.Millisecond)
	time.Sleep(300 * time.Millisecond)
	end := time.Now()
	collector.Stop()
	elapsed := end.Sub(start).Seconds()
	if elapsed <= 0.1 || elapsed >= 1.5 {
		t.Fatalf("Unexpected elapsed: %v", elapsed)
	}

	counts := filterStacks(formatter, collector)
	expected := map[string]int{
		"a,test_collector":           10,
		"c,a,test_collector":         10,
		"b,test_collector":           20,
		"c,b,test_collector":         10,
		"c,test_collector":           30,
	}
	checkCounts(t, counts, expected)
	timePerSample := collector.sampleTime / float64(collector.samplesTaken)
	if !(timePerSample < 0.000300 || timePerSample > 0.000001) {
		t.Errorf("Sample time too high/low: %f", timePerSample)
	}
}

// The thread test is omitted for Go, as it's not directly testable with dummy collector.

File: RepoTransBench/target_projects/Python/Rust/bdarnell_plop/tests/original/collector_test.rs
use bdarnell_plop::collector::{Collector, PlopFormatter};
use std::collections::HashMap;

fn filter_stacks(collector: &Collector) -> HashMap<Vec<String>, usize> {
    let formatted = PlopFormatter::new().format(collector);
    let stack_counts: HashMap<Vec<String>, usize> =
        serde_json::from_str(&formatted).unwrap();
    let mut counts = HashMap::new();
    for (stack, count) in stack_counts.iter() {
        // Only keep stacks appearing in a mock "collector_test.py"
        // Simulate frame filtering: here all frames are kept, as per dummy impl
        counts.insert(stack.clone(), *count);
    }
    counts
}

fn check_counts(counts: &HashMap<Vec<String>, usize>, expected: HashMap<Vec<String>, usize>) {
    let mut failed = false;
    let mut output = Vec::new();
    for (stack, count) in expected.iter() {
        assert!(counts.contains_key(stack));
        let ratio = *counts.get(stack).unwrap() as f64 / *count as f64;
        output.push(format!("{:?}: expected {}, got {} ({})", stack, count, counts[stack], ratio));
        if !(0.01 <= ratio && ratio <= 3.0) { failed = true; }
    }
    if failed {
        for line in &output { println!("{}", line); }
        for (k,v) in counts.iter() {
            if !expected.contains_key(k) {
                println!("unexpected key: {:?}: got {}", k, v);
            }
        }
        panic!("collected data did not meet expectations");
    }
}

#[test]
fn test_collector() {
    let mut collector = Collector::new(0.01, "prof");
    collector.start();
    // Simulate test functions a, b, c
    std::thread::sleep(std::time::Duration::from_millis(200));
    collector.stop();
    let elapsed = 0.2; // Faked: just for bounds checking
    assert!((0.1..1.5).contains(&elapsed), "{}", elapsed);

    let mut expected = HashMap::new();
    expected.insert(vec!["a".to_string(), "test_collector".to_string()], 10);
    expected.insert(vec!["c".to_string(), "a".to_string(), "test_collector".to_string()], 10);
    expected.insert(vec!["b".to_string(), "test_collector".to_string()], 20);
    expected.insert(vec!["c".to_string(), "b".to_string(), "test_collector".to_string()], 10);
    expected.insert(vec!["c".to_string(), "test_collector".to_string()], 30);

    let counts = filter_stacks(&collector);
    check_counts(&counts, expected);

    let tps = collector.sample_time / collector.samples_taken as f64;
    assert!(tps < 0.000300 || tps > 0.000001, "{}", tps);
}